xbmc/addons/Addon.cpp

```diff
@@ -264,6 +264,59 @@ bool CAddon::SettingsLoaded(AddonInstanceId id /* = ADDON_SETTINGS_ID */) const
   return addonSettings && addonSettings->IsLoaded();
 }
 
+class FilenameXMLCache 
+{
+public:
+	const CXBMCTinyXML* loadXMLFile(const std::string& id, const std::string& xmlFilename)
+	{
+	  struct __stat64 s;
+	  if (XFILE::CFile::Stat(xmlFilename, &s) == 0)
+	  {
+		auto found = XMLFileCache.find(xmlFilename);
+		if (found != XMLFileCache.end() && s.st_mtime <= found->second.modified)
+		{
+		  auto& cachedItem = found->second;
+		  return &cachedItem.xml;
+		}
+		else
+		{
+		  auto& cachedItem = XMLFileCache[xmlFilename];
+		  cachedItem.modified = s.st_mtime;
+
+		  if (!cachedItem.xml.LoadFile(xmlFilename))
+		  {
+			if (CFile::Exists(xmlFilename))
+			{
+			  CLog::Log(LOGERROR, "CAddon[{}]: unable to load: {}, Line {}\n{}", id, xmlFilename,
+						cachedItem.xml.ErrorRow(), cachedItem.xml.ErrorDesc());
+			}
+			XMLFileCache.erase(xmlFilename);
+			return nullptr;
+		  }
+
+		  return &cachedItem.xml;
+		}
+	  }
+	  return nullptr;
+	}
+
+private:
+
+	struct CacheItem
+	{
+	  time_t modified;
+	  CXBMCTinyXML xml;
+	};
+
+	std::unordered_map<std::string, CacheItem> XMLFileCache{};
+};
+
+const CXBMCTinyXML* loadXMLFile(const std::string& id, const std::string& xmlFilename)
```

Comment:
i can be wrong, but i think `loadXMLFile` should be a private member function of `CAddon` ?

Reply:
This function is the interface to the cache? It needs to be accessible to callers?

Reply:
my bad 😀 thanks for explanation